modules/core/src/mathfuncs.cpp

```diff
@@ -1600,50 +1608,437 @@ static bool ocl_patchNaNs( InputOutputArray _a, float value )
+            {
+                v_int32 v_src = vx_load(tptr + j);
+                v_int32 v_cmp_mask = v_mask2 < (v_src & v_mask1);
+                v_int32 v_dst = v_select(v_cmp_mask, v_val, v_src);
+                v_store(tptr + j, v_dst);
+            }
+            vx_cleanup();
+#endif
+
+            for (; j < len; j++)
+                if ((tptr[j] & 0x7fffffff) > 0x7f800000)
+                    tptr[j] = val.i;
+        }
+    }
+    else if (_a.depth() == CV_64F)
     {
-        int* tptr = ptrs[0];
-        size_t j = 0;
+        Cv64suf val;
+        val.f = _val;
+
+#if CV_SIMD
+        v_int64 v_mnt_mask = vx_setall_s64(0x000FFFFFFFFFFFFF), v_exp_mask = vx_setall_s64(0x7FF0000000000000);
+        v_int64 v_val = vx_setall_s64(val.i);
+#endif
+
+        for (size_t i = 0; i < it.nplanes; i++, ++it)
+        {
+            int64* tptr = (int64*)ptrs[0];
+            size_t j = 0;
+
+#if CV_SIMD
+            size_t cWidth = (size_t)v_int64::nlanes;
+            for (; j + cWidth <= len; j += cWidth)
+            {
+                v_int64 v_src = vx_load(tptr + j);
+                v_int64 v_isnan = ((v_src & v_exp_mask) == v_exp_mask) & ((v_src & v_mnt_mask) != vx_setzero_s64());
+                v_int64 v_dst = (v_isnan & v_val) | ((~v_isnan) & v_src);
+                v_store(tptr + j, v_dst);
+            }
+            vx_cleanup();
+#endif
+
+            for (; j < len; j++)
+                if ((tptr[j] & 0x7FFFFFFFFFFFFFFF) > 0x7FF0000000000000)
+                    tptr[j] = val.i;
+        }
+    }
+}
 
 #if CV_SIMD
-        size_t cWidth = (size_t)v_int32::nlanes;
-        for ( ; j + cWidth <= len; j += cWidth)
+template <typename _Tp, int cn>
+int nanMaskSIMD_(const _Tp *src, uchar *dst, size_t total, bool maskNans, bool maskInfs, bool maskAll, bool invert);
+
+//TODO: make true SIMD code instead
+template <>
+int nanMaskSIMD_<float, 1>(const float *src, uchar *dst, size_t total, bool maskNans, bool maskInfs, bool maskAll, bool invert)
+{
+    const int vsize = v_float32::nlanes;
+    int i = 0;
+    for(; i <= (int)total - vsize; i += vsize )
+    {
+        for (int j = 0; j < vsize; j++)
         {
-            v_int32 v_src = vx_load(tptr + j);
-            v_int32 v_cmp_mask = v_mask2 < (v_src & v_mask1);
-            v_int32 v_dst = v_select(v_cmp_mask, v_val, v_src);
-            v_store(tptr + j, v_dst);
+            bool nan = maskAll ? true : false;
+
+            float val = src[i + j];
+            bool v = (maskNans && cvIsNaN(val)) || (maskInfs && cvIsInf(val));
+            if (maskAll)
+                nan = nan && v;
+            else
+                nan = nan || v;
+
+            nan = invert ? !nan : nan;
+            dst[i + j] = nan ? 255 : 0;
         }
-        vx_cleanup();
+    }
+
+    return i;
+}
+
+//TODO: make true SIMD code instead
+template <>
+int nanMaskSIMD_<double, 1>(const double *src, uchar *dst, size_t total, bool maskNans, bool maskInfs, bool maskAll, bool invert)
```

Comment:
I propose not duplicate the implementation many times, but stay with template version only for now. Current batch of function copies does not make sense. Template specializations may be added in any moment without existing code modifications.

Reply:
This is just a boilerplate code for further SIMD implementations. Will be replace by a real code soon.